bpo-41877: Check for misspelled speccing arguments

vabr-g opened the pull request:
patch, patch.object and create_autospec silently ignore misspelled
arguments such as autospect, auto_spec and set_spec. This can lead
to tests failing to check what they are supposed to check.

This change adds a check causing a RuntimeError if the above
functions get any of the above misspellings as arguments. It also
adds a new argument, "unsafe", which can be set to True to disable
this check.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-41877](https://bugs.python.org/issue41877) -->
https://bugs.python.org/issue41877
<!-- /issue-number -->

vabr-g committed:
bpo-41877: Check for misspelled speccing arguments

patch, patch.object and create_autospec silently ignore misspelled
arguments such as autospect, auto_spec and set_spec. This can lead
to tests failing to check what they are supposed to check.

This change adds a check causing a RuntimeError if the above
functions get any of the above misspellings as arguments. It also
adds a new argument, "unsafe", which can be set to True to disable
this check.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

gpshead commented on Lib/unittest/mock.py:
```diff
@@ -1242,14 +1242,25 @@ def _importer(target):
+    for typo in typos:
+        if typo in kwargs_to_check:
+            raise RuntimeError(
+                f"{typo} might be a typo; use unsafe=True if this is intended"
```
Change `{typo}` to `{typo!r}` so that it uses a repr to gain quotes?  (and double check that it has the desired effect)

It occurs to me that I didn't do that in the edit on the earlier PR updating the error message, so also make this change to that error message as well?

vabr-g replied:
Thanks, good point. I changed both places. These are sample error messages:

> RuntimeError: 'autospect' might be a typo; use unsafe=True if this is intended

> AttributeError: 'assert_foo_call' is not a valid assertion. Use a spec for the mock if 'assert_foo_call' is meant to be an attribute.

gpshead commented on Lib/unittest/mock.py:
```diff
@@ -1242,14 +1242,25 @@ def _importer(target):
     def __init__(
             self, getter, attribute, new, spec, create,
-            spec_set, autospec, new_callable, kwargs
+            spec_set, autospec, new_callable, unsafe, kwargs
```
code style wise, this is already a very long list of arguments.  when adding another one, I'd add it as a keyword only argument at the end.

`, kwargs, *, unsafe):`

vabr-g replied:
Thanks for teaching me about keyword-only arguments. This also allowed me to drop some unnecessary changes done previously just to supply the positional argument correctly.

gpshead commented on Lib/unittest/mock.py:
```diff
@@ -2567,8 +2584,8 @@ def call_list(self):
 
-def create_autospec(spec, spec_set=False, instance=False, _parent=None,
-                    _name=None, **kwargs):
+def create_autospec(spec, spec_set=False, instance=False, unsafe=False,
```
as unsafe is being added to the API here, perhaps make it keyword only?

vabr-g replied:
Done.

vabr-g committed:
Make `unsafe` a kw-only arg

Also revert changes forced by `unsafe` being positional,
and add "!r" to format specifiers in added error messages.

vabr-g committed:
Make `unsafe` a kw-only arg

Also revert changes forced by `unsafe` being positional,
and add "!r" to format specifiers in added error messages.

vabr-g committed:
Fix indenting.

vabr-g committed:
Merge branch 'issue-41877-arguments-check' of github.com:vabr-g/cpython into issue-41877-arguments-check

gpshead commented on Lib/unittest/mock.py:
```diff
@@ -1242,14 +1242,25 @@ def _importer(target):
+# _check_spec_arg_typos takes kwargs from commands like patch and checks that
+# they don't contain common misspellings of arguments related to autospeccing.
+def _check_spec_arg_typos(kwargs_to_check):
+    typos = ["autospect", "auto_spec", "set_spec"]
```
as this is a constant I suggest using a () tuple instead of [] list.

vabr-g replied:
Done.

gpshead commented on Lib/unittest/test/testmock/testmock.py:
```diff
@@ -38,6 +38,12 @@ def cmeth(cls, a, b, c, d=None): pass
     def smeth(a, b, c, d=None): pass
 
 
+class Typos(object):
```
No need to inherit from (object) anymore, that's py2 style. :)

vabr-g replied:
Done.

gpshead commented on Lib/unittest/test/testmock/testmock.py:
```diff
@@ -2175,6 +2181,52 @@ def __init__(self):
         self.assertEqual(obj.obj_with_bool_func.__bool__.call_count, 0)
 
+    def test_misspelled_arguments(self):
+        class Foo(object):
```
same here, no need for (object)

vabr-g replied:
Done.

gpshead reviewed (commented):
all my remaining comments are minor

vabr-g committed:
List -> tuple; drop object inheritance.
